crates/nargo/src/ops/foreign_calls.rs

```diff
@@ -71,69 +64,88 @@ impl ForeignCall {
-        let (abi_type, input_values) = fetch_abi_type(foreign_call_inputs)?;
+        let (is_fmt_str, foreign_call_inputs) =
+            foreign_call_inputs.split_last().ok_or(ForeignCallError::MissingForeignCallInputs)?;
+
+        let output_string = if is_fmt_str[0].to_field().is_one() {
+            convert_fmt_string_inputs(foreign_call_inputs)?
+        } else {
+            convert_string_inputs(foreign_call_inputs)?
+        };
+        println!("{output_string}");
+        Ok(())
+    }
+}
+
+fn convert_string_inputs(foreign_call_inputs: &[Vec<Value>]) -> Result<String, ForeignCallError> {
+    // Fetch the abi type from the foreign call input
+    // The remaining input values should hold what is to be printed
+    let (abi_type_as_values, input_values) =
+        foreign_call_inputs.split_last().ok_or(ForeignCallError::MissingForeignCallInputs)?;
+    let abi_type = fetch_abi_type(abi_type_as_values)?;
 
-        // We must use a flat map here as each value in a struct will be in a separate input value
-        let mut input_values_as_fields =
-            input_values.iter().flat_map(|values| vecmap(values, |value| value.to_field()));
-        let decoded_value = decode_value(&mut input_values_as_fields, &abi_type)?;
+    // We must use a flat map here as each value in a struct will be in a separate input value
+    let mut input_values_as_fields =
+        input_values.iter().flat_map(|values| vecmap(values, |value| value.to_field()));
 
-        let json_value = JsonTypes::try_from_input_value(&decoded_value, &abi_type)?;
-        let output_string = serde_json::to_string_pretty(&json_value)
-            .map_err(|err| ForeignCallError::InputParserError(err.into()))?;
+    let decoded_value = decode_value(&mut input_values_as_fields, &abi_type)?;
+    let json_value = JsonTypes::try_from_input_value(&decoded_value, &abi_type)?;
 
-        println!("{output_string}");
-        Ok(())
+    serde_json::to_string_pretty(&json_value)
+        .map_err(|err| ForeignCallError::InputParserError(err.into()))
+}
+
+fn convert_fmt_string_inputs(
+    foreign_call_inputs: &[Vec<Value>],
+) -> Result<String, ForeignCallError> {
+    let (message_as_values, input_and_abi_values) =
+        foreign_call_inputs.split_first().ok_or(ForeignCallError::MissingForeignCallInputs)?;
+
+    let message_as_fields = vecmap(message_as_values, |value| value.to_field());
+    let message_as_string = decode_string_value(&message_as_fields);
+
+    let (num_values, input_and_abi_values) =
+        input_and_abi_values.split_first().ok_or(ForeignCallError::MissingForeignCallInputs)?;
+
+    let mut output_strings = Vec::new();
+    let num_values = num_values[0].to_field().to_u128() as usize;
+
+    let mut abi_types = Vec::new();
+    for i in (input_and_abi_values.len() - num_values)..input_and_abi_values.len() {
+        let abi_type = fetch_abi_type(&input_and_abi_values[i])?;
+        abi_types.push(abi_type);
     }
 
```

Comment:
nit: Perhaps using .skip makes this easier to read, pseudocode:

```rust
    let skip_amount = input_and_abi_values.len() - num_values;
    for i in (0..input_and_abi_values.len()).skip(skip_amount) {
        let abi_type = fetch_abi_type(&input_and_abi_values[i])?;
        abi_types.push(abi_type);
    }
```

What do you think?

Reply:
I switched to `skip`. But I did this: 
```
for abi_values in input_and_abi_values.iter().skip(input_and_abi_values.len() - num_values) {
```